Sources/WalletConnectSign/Engine/Controller/ApproveEngine.swift

```diff
@@ -0,0 +1,309 @@
+    
+    private var publishers = Set<AnyCancellable>()
+
+    init(
+        networkingInteractor: NetworkInteracting,
+        proposalPayloadsStore: CodableStore<WCRequestSubscriptionPayload>,
+        sessionToPairingTopic: CodableStore<String>,
+        metadata: AppMetadata,
+        kms: KeyManagementServiceProtocol,
+        logger: ConsoleLogging,
+        pairingStore: WCPairingStorage,
+        sessionStore: WCSessionStorage
+    ) {
+        self.networkingInteractor = networkingInteractor
+        self.proposalPayloadsStore = proposalPayloadsStore
+        self.sessionToPairingTopic = sessionToPairingTopic
+        self.metadata = metadata
+        self.kms = kms
+        self.logger = logger
+        self.pairingStore = pairingStore
+        self.sessionStore = sessionStore
+        
+        setupNetworkingSubscriptions()
+    }
+    
+    func approveProposal(proposerPubKey: String, validating sessionNamespaces: [String: SessionNamespace]) throws {
+        let payload = try proposalPayloadsStore.get(key: proposerPubKey)
+
+        guard let payload = payload, case .sessionPropose(let proposal) = payload.wcRequest.params else {
+            throw Error.wrongRequestParams
+        }
+
+        proposalPayloadsStore.delete(forKey: proposerPubKey)
+        
+        try Namespace.validate(sessionNamespaces)
+        try Namespace.validateApproved(sessionNamespaces, against: proposal.requiredNamespaces)
+        
+        let selfPublicKey = try kms.createX25519KeyPair()
+
+        guard let agreementKey = try? kms.performKeyAgreement(
+            selfPublicKey: selfPublicKey,
+            peerPublicKey: proposal.proposer.publicKey
+        ) else { throw Error.agreementMissingOrInvalid }
+
+        // TODO: Extend pairing
+        let sessionTopic = agreementKey.derivedTopic()
+        try kms.setAgreementSecret(agreementKey, topic: sessionTopic)
+
+        guard let relay = proposal.relays.first else {
+            throw Error.relayNotFound
+        }
+
+        let proposeResponse = SessionType.ProposeResponse(relay: relay, responderPublicKey: selfPublicKey.hexRepresentation)
+        let response = JSONRPCResponse<AnyCodable>(id: payload.wcRequest.id, result: AnyCodable(proposeResponse))
+        networkingInteractor.respond(topic: payload.topic, response: .response(response)) { _ in }
+
+        try settle(topic: sessionTopic, proposal: proposal, namespaces: sessionNamespaces)
+    }
+    
+    func reject(proposerPubKey: String, reason: ReasonCode) throws {
+        guard let payload = try proposalPayloadsStore.get(key: proposerPubKey) else {
+            throw Error.proposalPayloadsNotFound
+        }
+        proposalPayloadsStore.delete(forKey: proposerPubKey)
+        networkingInteractor.respondError(for: payload, reason: reason)
+        // TODO: Delete pairing if inactive
+    }
+    
+    func settle(topic: String, proposal: SessionProposal, namespaces: [String: SessionNamespace]) throws {
```

Comment:
I wonder if something can make the settlement fail after the proposal is responded with a success, and if this can cause any problems. Something that we can think about in the future.